rats-devtools/src/python/rats/annotations/_functions.py

```diff
@@ -0,0 +1,102 @@
+# type: ignore
+from __future__ import annotations
+
+from collections import defaultdict
+from collections.abc import Callable
+from functools import cache
+from typing import Any, NamedTuple, ParamSpec, T
+
+from rats.annotations._groups import GroupAnnotations, T_GroupType
+
+
+class AnnotationsContainer(NamedTuple):
+    """
+    Holds metadata about the annotated service provider.
+
+    Loosely inspired by: https://peps.python.org/pep-3107/.
+    """
+
+    annotations: tuple[GroupAnnotations[...], ...]
+
+    @staticmethod
+    def empty() -> AnnotationsContainer:
+        return AnnotationsContainer(annotations=())
+
+    def with_group(
+        self,
+        namespace: str,
+        group_id: T_GroupType,
+    ) -> AnnotationsContainer:
+        return AnnotationsContainer(
+            annotations=tuple([x for x in self.with_namespace(namespace) if group_id in x.groups]),
+        )
+
+    def with_namespace(
+        self,
+        namespace: str,
+    ) -> AnnotationsContainer:
+        return AnnotationsContainer(
+            annotations=tuple([x for x in self.annotations if x.namespace == namespace]),
+        )
+
+
+class AnnotationsBuilder:
+    _group_ids: dict[str, set[Any]]
+
+    def __init__(self) -> None:
+        self._group_ids = defaultdict(set)
+
+    def add(self, namespace: str, group_id: NamedTuple) -> None:
+        self._group_ids[namespace].add(group_id)
+
+    def make(self, name: str) -> AnnotationsContainer:
+        return AnnotationsContainer(
+            annotations=tuple(
+                [
+                    GroupAnnotations[Any](name=name, namespace=namespace, groups=tuple(groups))
+                    for namespace, groups in self._group_ids.items()
+                ]
+            ),
+        )
+
+
+AnnotationDecorator = Callable[[T], T]
```

Comment:
Found a couple of typing bugs with this that I've solved in [b0475d0](https://github.com/microsoft/rats/pull/161/commits/b0475d0158d34fc0efd25db9da30221a895b7aad), if you want to cherry-pick that (or we can sort it out once we figure out the blend of ideas in the other thread)